Downloader/SPUDownloader.m

```diff
@@ -111,12 +111,25 @@ - (void)enableAutomaticTermination
     }
 }
 
-- (void)cancelDownload
+- (NSString *)rootPersistentDownloadCachePathForBundleIdentifier:(NSString *)bundleIdentifier
 {
-    [self cleanup];
+    return [[SPULocalCacheDirectory cachePathForBundleIdentifier:bundleIdentifier] stringByAppendingPathComponent:@"PersistentDownloads"];
 }
 
-- (void)cleanup
+- (void)removeDownloadDirectory:(NSString *)directoryName bundleIdentifier:(NSString *)bundleIdentifier
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        NSString *rootPersistentDownloadCachePath = [self rootPersistentDownloadCachePathForBundleIdentifier:bundleIdentifier];
+        if (rootPersistentDownloadCachePath != nil) {
+            NSString *sanitizedDirectoryName = directoryName.lastPathComponent;
+            NSString *tempDir = [rootPersistentDownloadCachePath stringByAppendingPathComponent:sanitizedDirectoryName];
```

Comment:
This seems like duplicated code. If that dir is expected to exist, there must be another function somewhere that builds the path this way.

Reply:
Yeah.. there's two reasons: 1. I don't want to send full path across XPC so I let this class compute most that it can. 2. The downloader instance and update session that creates the temp directory is not necessarily the same as the one used to clear the download later, so can't "save" it in the downloader or XPC Service.

Alternatively I could send/pass full path and have downloader validate sub-components but seems less safe.